### PR TITLE
Initial signer/extension communication through WebSockets + JSON-RPC

### PR DESCRIPTION
**Pull request by Mrtenz**

### CODE REVIEW

File: packages/extension/src/relay/handlers/content.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      return;
+    }
+
+    const response = await sendMessage(event.data);
+    if (
+      !is(response, RelayResponseStruct) ||
+      response.id !== event.data.id ||
+      response.target !== RelayTarget.Content
+    ) {
+      return;
```
Comment (FrederikBolding) [inline packages/extension/src/relay/handlers/content.ts:20]: Can we test the error state?

File: packages/extension/src/store/jsonrpc.slice.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+const initialState: JsonRpcState = {
+  network: {
+    providers: ['https://api.mycryptoapi.com/eth'],
+    chainId: 1
+  }
+};
+
+const sliceName = 'jsonrpc';
+
+const slice: Slice<JsonRpcState> = createSlice({
```
Comment (FrederikBolding) [inline packages/extension/src/store/jsonrpc.slice.ts:20]: This slice looks to not be in use?

File: packages/extension/src/store/sagas.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { all } from 'redux-saga/effects';
+
+import { socketsSaga } from './sockets.slice';
+
+export default function* rootSaga(/* ipc: ReduxIPC */): Generator {
```
Comment (FrederikBolding) [inline packages/extension/src/store/sagas.ts:5]: Can we remove the parameter comment here?

File: packages/extension/src/store/sockets.slice.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    takeEvery(handleRequest.type, handleRequestWorker),
+    takeLatest(setConnected.type, setConnectedWorker)
+  ]);
+}
+
+/**
+ * Creates a new random private key to be used for signing requests.
+ */
+export function* createPrivateKeyWorker() {
+  const privateKey: string = yield call(createRandomPrivateKey);
```
Comment (FrederikBolding) [inline packages/extension/src/store/sockets.slice.ts:76]: Can we test this saga?

File: packages/extension/src/types/relay.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import type { Infer } from 'superstruct';
+import { any, array, number, object, optional, pattern, string, union, unknown } from 'superstruct';
+
+export enum RelayTarget {
+  Page = 'Page',
+  Content = 'Content',
+  Background = 'Background',
+  External = 'External'
```
Comment (FrederikBolding) [inline packages/extension/src/types/relay.ts:8]: What's this needed for?

File: packages/extension/src/relay/background.test.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
+import { generateUUID } from '@signer/common';
+import { chrome } from 'jest-chrome';
+
+import { RelayTarget } from '../types';
+import { createBackgroundMessageSender } from './background';
+
+const sleep = () => new Promise((resolve) => setTimeout(resolve, 1));
```
Comment (FrederikBolding) [inline packages/extension/src/relay/background.test.ts:7]: Can we share this code between the files that use it?

File: packages/extension/src/store/sockets.slice.test.ts [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+      .fork(socketRequestWorker, instance);
+
+    callback({ data: JSON.stringify({ foo: 'bar' }) } as MessageEvent);
+
+    saga.put(message({ foo: 'bar' }));
+
+    await promise;
+  });
+
+  it('reconnects if the connection is closed', async () => {
```
Comment (FrederikBolding) [inline packages/extension/src/store/sockets.slice.test.ts:134]: How does this test work? 😆

File: packages/extension/src/store/sockets.slice.ts [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+  try {
+    while (true) {
+      try {
+        const action: MessageEvent = yield take(channel);
+        const json = JSON.parse(action.data);
+
+        yield put(message(json));
+      } catch (error) {
+        yield put(setConnected(false));
+        channel.close();
```
Comment (FrederikBolding) [inline packages/extension/src/store/sockets.slice.ts:167]: Any idea why these aren't these covered?

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

File: packages/extension/src/relay/handlers/content.ts (hunk 1 above)
Comment (Mrtenz) [reply packages/extension/src/relay/handlers/content.ts:20]: What do you mean? This is not an error state

File: packages/extension/src/utils/inject.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+export const injectScript = (): Promise<void> => {
+  return new Promise((resolve) => {
+    const script = document.createElement('script');
+    script.src = chrome.runtime.getURL('lib/page.js');
+    script.addEventListener('load', () => {
+      script.remove();
+      resolve();
```
[the remark this answers is not in the document]
Comment (Mrtenz) [reply packages/extension/src/utils/inject.ts:7]: I have no idea how we would test this.

File: packages/extension/src/store/jsonrpc.slice.ts (hunk 2 above)
Comment (Mrtenz) [reply packages/extension/src/store/jsonrpc.slice.ts:20]: It's not in use right now, but will be needed for handling requests to an actual JSON-RPC node.

File: packages/extension/src/types/relay.ts (hunk 5 above)
Comment (Mrtenz) [reply packages/extension/src/types/relay.ts:8]: Two things:
- Depending on how messages are sent across browser extensions, they may not be fully isolated from other extensions. This basically checks if messages sent are coming from a part of the extension.
- This ensures that messages aren't replayed twice. Calling `window.postMessage` in the content script will also trigger the event listener in the content script.

File: packages/extension/src/store/sockets.slice.test.ts (hunk 7 above)
Comment (Mrtenz) [reply packages/extension/src/store/sockets.slice.test.ts:134]: It gets the `close` event callback (`addEventListener.mock.calls[2][1]`) and calls it manually.

File: packages/extension/src/utils/inject.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply packages/extension/src/utils/inject.ts:7]: 👍

File: packages/extension/src/relay/handlers/content.ts (hunk 1 above)
Comment (FrederikBolding) [reply packages/extension/src/relay/handlers/content.ts:20]: Sorry, can we test this edge case? I just mean that this line isn't covered 😄

File: packages/extension/src/types/relay.ts (hunk 5 above)
Comment (FrederikBolding) [reply packages/extension/src/types/relay.ts:8]: Sorry, I specifically meant `External`. I couldn't find it used anywhere?

**Review by FrederikBolding — APPROVED** [review]
